app/code/Magento/SalesRule/Plugin/CouponUsagesIncrement.php

```diff
@@ -34,14 +36,16 @@ public function __construct(
      * Increments number of coupon usages after placing order.
      *
      * @param OrderService $subject
-     * @param OrderInterface $result
-     * @return OrderInterface
+     * @param OrderInterface $order
+     * @return array
+     * @throws CouponUsageExceeded
+     * @throws LocalizedException
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
      */
-    public function afterPlace(OrderService $subject, OrderInterface $result): OrderInterface
+    public function beforePlace(OrderService $subject, OrderInterface $order): array
```

Comment:
I don't think this plugin should happen before the order placement, if any other problem happens to the order place flow it will cause problems where coupons will be used and there are no orders placed.

Reply:
Hi @gabrieldagama , 

Please see the comment I've added on: app/code/Magento/Checkout/Api/PaymentInformationManagementInterface.php

Your concerns are covered by the way coupon management was implemented originally and I wanted to keep the changes as low as posible:
\Magento\SalesRule\Observer\DecrementCouponUsageObserver
\Magento\SalesRule\Plugin\CouponUsagesDecrement

We need to:

-  increment the voucher usage before placing the order because voucher processing is not inside order save transaction 
- break the order placement process in case the voucher is already used at the moment of beginning order place action, to avoid saving the order
- show customer the error and the new total
- **in case of error in order placement the coupon usage decrement is covered by original behavior**